[NVPTX] Support fence instruction

gonzalobg opened the pull request.

gonzalobg committed:
[NVPTX] Support fence instruction

gonzalobg committed:
Fixup: use INT_MEMBAR_SYS instead of redefining the membar.sys opcode

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3908,3 +3908,29 @@ def : Pat <
+def : Pat<(atomic_fence (i64 7), (i64 1)), (atomic_thread_fence_seq_cst_sys)>, // seq_cst(7) sys(1)
+      Requires<[hasPTX<60>, hasSM<70>]>;
+
+def : Pat<(atomic_fence (i64 4), (i64 1)), (INT_MEMBAR_SYS)>; // acquire(4) sys(1)
```
These patterns appear to match the same patterns as above, so one of them is probably not needed. Unless I miss something.

gonzalobg replied:
These patterns do match the same patterns as above, but don't have the `Requires<[hasPTX<60>, hasSM<70>]>`, that is, when the above patterns fail to match due to that `Requires` failing, these patterns do match and cover the cases of PTX < 60 or SM < 70.

That was at least my intent, maybe I got that wrong. 

The  `Requires<[hasPTX<60>, hasSM<70>]>` on the `def`s doesn't seem to do anything, but I left there for documentation purposes.

Artem-B replied:
Inded, I've missed that part. It may be useful to add a commenr that pre-sm60 have to resort to `membar.sys` hammer.

gonzalobg replied:
```suggestion

// If PTX<60 or SM<70, we fall back to MEMBAR:
def : Pat<(atomic_fence (i64 4), (i64 1)), (INT_MEMBAR_SYS)>; // acquire(4) sys(1)
```

gonzalobg committed:
[NVPTX] Add comment to clarify pre sm70 behavior

gonzalobg commented:
@Artem-B I don't have permission to merge; i've rebased and all tests are green (I rebased a few days ago, and some tests failed spuriously). Could you merge this for me? Thank you!
